src/sentry/static/sentry/app/utils/apm.jsx

```diff
@@ -30,12 +30,72 @@ function startTransaction() {
   finishTransaction(5000);
 }
 
+const requests = new Map([]);
+const renders = new Map([]);
 let flushTransactionTimeout = undefined;
-function finishTransaction(delay) {
+let interruptFlush = false;
+export function finishTransaction(delay) {
   if (flushTransactionTimeout) {
     clearTimeout(flushTransactionTimeout);
   }
-  flushTransactionTimeout = setTimeout(() => Sentry.finishSpan(), delay || 5000);
+  if (
+    Array.from(requests).find(([, active]) => active) ||
+    Array.from(renders).find(([, active]) => active)
+  ) {
+    clearTimeout(flushTransactionTimeout);
+  }
+  flushTransactionTimeout = setTimeout(() => {
+    Sentry.finishSpan();
+  }, delay || 5000);
+}
+
+export function startRequest(id) {
+  // if flush is active, stop it
+  if (flushTransactionTimeout) {
+    clearTimeout(flushTransactionTimeout);
+    interruptFlush = true;
+  }
+
+  requests.set(id, true);
+}
+export function finishRequest(id) {
+  requests.set(id, false);
+
+  if (
+    interruptFlush &&
+    !Array.from(requests).find(([, active]) => active) &&
+    !Array.from(renders).find(([, active]) => active)
+  ) {
+    finishTransaction(1);
```

Comment:
You can use `finishTransaction(0)` when you change it's implementation to `finishTransaction(delay = 5000)` and `  }, delay || 5000);` to `  }, delay);`